feat(Provenance): Add `DirectoryProvenance` as a `LocalProvenance`

pepper-jk opened the pull request:
In contrast to the previously added `RemoteProvenance` stands the `LocalProvenance`, which has no remote source, but instead references a local input of some kind.
The `DirectoryProvenance` references a local project directory, which is lacking supported (remote) version control. It is defined by its local directory path only.

Since ORT needs further refactoring until `DirectoryProvenance` can be fully utilized, it the new class can not be used right now. However the presence of a new `KnownProvenance` class, results in `when` conditional cases not being exhaustive anymore.

To circumvent this issue, the following changes were made:
1. Whereever possible `RemoteProvenance` is used as parameter type instead of `KnownProvenance`.
2. When necessary `KnownProvenance`s are cast to `RemoteProvenance`.
3. If `Provenance` is expected, `DirectoryProvenance` is handled like `UnknownProvenance`.

The excaption being `hash` and `storageKey`, which both default to an empty string. However, since the rest of the code does not handle `DirectoryProvenance`, this should not become an issue.

For more context on the new Provenance Hierarchy, see: https://github.com/oss-review-toolkit/ort/issues/8803#issuecomment-2236958430

Signed-off-by: Jens Keim <jens.keim@forvia.com>

pepper-jk committed:
feat(Provenance): Add `DirectoryProvenance` as a `LocalProvenance`

In contrast to the previously added `RemoteProvenance` stands the
`LocalProvenance`, which has no remote source, but instead references
a local input of some kind.
The `DirectoryProvenance` references a local project directory,
which is lacking supported (remote) version control.
It is defined by its canonical path only.

Since ORT needs further refactoring until `DirectoryProvenance` can
be fully utilized, the new class can not be used right now.
However the presence of a new `KnownProvenance` class results in
`when` conditional cases not being exhaustive anymore.

To circumvent this issue, the following changes were made:
1. Wherever possible `RemoteProvenance` is used as parameter type
   instead of `KnownProvenance`.
2. When necessary `KnownProvenance`s are cast to `RemoteProvenance`.
3. If `Provenance` is expected, `DirectoryProvenance` is handled
   like `UnknownProvenance`.

For instances of `Package` the new default data structure should be
`RemoteProvenance`, as a `Package` by definition requires a remote.

The exception being `hash` and `storageKey`, which both required a
default value, which was set to the `canonicalPath`.
However, since the rest of the code does not handle
`DirectoryProvenance`, this should remain unused for now.

See [1] for more context on the new provenance hierarchy.

[1]: https://github.com/oss-review-toolkit/ort/issues/8803#issuecomment-2236958430

Signed-off-by: Jens Keim <jens.keim@forvia.com>

pepper-jk commented:
@sschuberth failing tests seem unrelated.
Could you have a look and maybe restart them?
A review would be awesome as well.

sschuberth commented:
> @sschuberth failing tests seem unrelated.

The "Unable to create proxy for sealed class" failures in `org.ossreviewtoolkit.scanner.ScannerTest` are real. I guess the problem is that nothing currently implements the `LocalProvenance` interface. So you probably have to add a `DirectoryProvenance` at the same time.

pepper-jk commented:
Thanks, I missed that. I'm already working on the `DirectoryProvenance` over at [provenance-directory](https://github.com/pepper-jk/ort/commits/provenance-directory/).
So it would be easy enough to append it here.

However, it is quite a large branch, since a lot of `when`s require a `DirectoryProvenance` case.
Most of which are for `KnownProvenance` input, so I can not default to `UnknownProvenance` behavior.

I started with reasonable implementation for the `DirectoryProvenance` behavior, but the scope of required does not seem to end. So I was considering switching to `RemoteProvenance` inputs to avoid the usage of `DirectoryProvenance` for now.

Any thoughts or other ideas how to keep the PR small?

sschuberth commented:
> Most of which are for `KnownProvenance` input, so I can not default to `UnknownProvenance` behavior.

Can't you "cheat" a bit by first limiting some / all `KnownProvenance` cases to `RemoteProvenance` (which semantically is no difference compared to before), and only later bit-by-bit extend respective code paths to also handle `LocalProvenance`?

pepper-jk commented:
> So I was considering switching to `RemoteProvenance` inputs to avoid the usage of `DirectoryProvenance` for now.

> Can't you "cheat" a bit by first limiting some / all `KnownProvenance` cases to `RemoteProvenance` (which semantically is no difference compared to before), and only later bit-by-bit extend respective code paths to also handle `LocalProvenance`?

Great to see we are on the same page here. I will do just that.

pepper-jk commented:
@sschuberth I added the `DirectoryProvenance` now and "cheated" the `when` cases.
Let me know what you think, if you have any reservations.

sschuberth commented:
Commit message:

> It is defined by its local directory path only.

We should probably say "absolute" / "canonical" / "real" path, and do the implementation accordingly.

> it the new class can not be used right now.

"the new class can not be used right now."

> However the presence of a new `KnownProvenance` class, results

"However, the presence of a new `KnownProvenance` class results"

> Whereever

"Wherever"

> excaption

"exception"

sschuberth commented on model/src/main/kotlin/utils/FileProvenanceFileStorage.kt:
```diff
@@ -81,6 +82,7 @@ private fun KnownProvenance.hash(): String {
     val key = when (this) {
         is ArtifactProvenance -> "${sourceArtifact.url}${sourceArtifact.hash.value}"
         is RepositoryProvenance -> "${vcsInfo.type}${vcsInfo.url}$resolvedRevision"
+        is DirectoryProvenance -> ""
```
I don't think it's a good idea to (even temporarily) return an empty string here. We could consider to use an SPDX package verification code on the directory here. However, that's potentially expensive to compute. Why not just use the `canonicalPath` property? That's at least better than the empty string.

sschuberth commented on model/src/main/kotlin/utils/PurlExtensions.kt:
```diff
@@ -101,6 +102,8 @@ fun Provenance.toPurlExtras(): PurlExtras =
             )
         }
 
+        is DirectoryProvenance -> PurlExtras()
+
```
I'd instead change the existing `is UnknownProvenance -> PurlExtras()` line to

    !is RemoteProvenance -> PurlExtras()

because again purls only make sense for something that has been published to a remote.

sschuberth commented on scanner/src/main/kotlin/provenance/ProvenanceDownloader.kt:
```diff
@@ -61,7 +61,7 @@ fun interface ProvenanceDownloader {
         // likely already cached if a path scanner wrapper is used.
 
-        val root = download(nestedProvenance.root)
+        val root = download(nestedProvenance.root as RemoteProvenance)
```
I'm not yet convinced about the casts here an in other places, but I current do not have the time to think about it more.

pepper-jk replied:
I understand your hesitation. I was not 100% happy with the change either.

I just added a commit 4b979057c04ba87245396ec2f78fae4612d76b68 to this PR, which removes the casts.
Unfortunately it also requires a lot of file changes.

Let me know which direction you prefer. Then I can either squash or remove the commit.

pepper-jk replied:
Hi @sschuberth. Any way we can move forward on this?

@oss-review-toolkit/kotlin-devs Maybe someone else could have a look, if you are busy.

sschuberth replied:
Thanks for the reminder. I'm reviewing it now as we speak.

pepper-jk committed:
refactor: Remove `RemoteProvenance` casts

Casts were removed from `FileListResolver`, `ProvenanceBasedFileStorage`,
`ProvenanceDownloader`, and `Scanner`.

In order to avoid casting `KnownProvenance` to `RemoteProvenance`,
a lot of parameters need to be changed to `RemoteProvenance`.

With the exception of `ProvenanceBasedFileStorage`, which now uses
`UnknownProvenance` exceptions just like `ProvenanceBasedPostgresStorage`.

Signed-off-by: Jens Keim <jens.keim@forvia.com>

pepper-jk commented on model/src/main/kotlin/Provenance.kt:
```diff
@@ -83,6 +87,23 @@ data class RepositoryProvenance(
+data class DirectoryProvenance(
+    val canonicalPath: File
+) : LocalProvenance {
+    init {
+        require(canonicalPath.isDirectory) { "The directory path must exist." }
+    }
+
```
This makes sense to me. Just need to keep it in mind once we actually scan the directory.

pepper-jk replied:
Done.

pepper-jk commented on model/src/main/kotlin/ProvenanceResolutionResult.kt:
```diff
@@ -37,7 +37,7 @@ data class ProvenanceResolutionResult(
      */
     @JsonInclude(JsonInclude.Include.NON_NULL)
-    val packageProvenance: KnownProvenance? = null,
+    val packageProvenance: RemoteProvenance? = null,
```
Sounds about right. I can add it to the commit message.

Would have to check if it is these are the only places, where we switch to `RemoteProvenance` tho before removing the previous explanation.

pepper-jk replied:
Added the remark to the first commit message.

pepper-jk commented on scanner/src/main/kotlin/storages/ProvenanceBasedPostgresStorage.kt:
```diff
@@ -86,6 +87,10 @@ class ProvenanceBasedPostgresStorage(
                 val query = table.selectAll()
 
+                if (provenance !is RemoteProvenance) {
+                    throw ScanStorageException("Scan result must have a known provenance, but it is $provenance.")
```
Yes, the code and the exception's text are technically not reflecting the same thing.

However, while the `DirectoryProvenance` is unused, the `RemoteProvenance` is practically the same as `KnownProvenance`. Similar to my comment above, this is a temporary change, which aims to avoid issues due unfinished class handling.

pepper-jk commented on model/src/main/kotlin/Provenance.kt:
```diff
@@ -83,6 +87,23 @@ data class RepositoryProvenance(
+
+    /**
+     * Return true if this provenance's directoryPath matches the package URL of the [package][pkg],
+     * as it contains the local file path for non-remote Provenances.
```
Done.

pepper-jk commented on model/src/main/kotlin/utils/FileProvenanceFileStorage.kt:
```diff
@@ -81,6 +82,7 @@ private fun KnownProvenance.hash(): String {
     val key = when (this) {
         is ArtifactProvenance -> "${sourceArtifact.url}${sourceArtifact.hash.value}"
         is RepositoryProvenance -> "${vcsInfo.type}${vcsInfo.url}$resolvedRevision"
+        is DirectoryProvenance -> "${canonicalPath.toPath()}${canonicalPath.hashCode()}"
```
Done.

pepper-jk commented on model/src/main/kotlin/utils/PostgresProvenanceFileStorage.kt:
```diff
@@ -121,4 +122,5 @@ private fun KnownProvenance.storageKey(): String =
         is ArtifactProvenance -> "source-artifact|${sourceArtifact.url}|${sourceArtifact.hash}"
         // The trailing "|" is kept for backward compatibility because there used to be an additional parameter.
         is RepositoryProvenance -> "vcs|${vcsInfo.type}|${vcsInfo.url}|$resolvedRevision|"
+        is DirectoryProvenance -> "directory|${canonicalPath.toPath()}|${canonicalPath.hashCode()}"
```
Done.

pepper-jk commented on model/src/main/kotlin/utils/PurlExtensions.kt:
```diff
@@ -101,7 +102,7 @@ fun Provenance.toPurlExtras(): PurlExtras =
         }
 
-        is UnknownProvenance -> PurlExtras()
+        !is RemoteProvenance -> PurlExtras()
```
Done.

pepper-jk commented on scanner/src/main/kotlin/storages/ProvenanceBasedPostgresStorage.kt:
```diff
@@ -137,7 +142,7 @@ class ProvenanceBasedPostgresStorage(
         requireEmptyVcsPath(provenance)
 
-        if (provenance !is KnownProvenance) {
+        if (provenance !is RemoteProvenance) {
```
Same. ^^

pepper-jk committed:
fix(Scanner): RemoteProvenance for ORT 66.0.1 (972e3f6)

Signed-off-by: Jens Keim <jens.keim@forvia.com>
